examples/learn-tensors-with-mcov/learn-tensors-with-mcov.py

```diff
@@ -0,0 +1,445 @@
+#   \Big(\mathbf{q}^{\widetilde{\otimes} l}_1\widetilde{\otimes}
+#   \mathbf{q}_2^{\widetilde{\otimes} (\lambda-l-1)}\widetilde{\otimes}
+#   \mathbf{q}_3\Big)_{\lambda\mu}.
+#
+# 4. :math:`\lambda`-Correction Term
+# """"""""""""""""""""""""""""""""""
+#
+# Highly symmetric environments can lead to all-zero vector spherical expansion
+# components, which in turn would yield all-zero tensor features.
+# To correct this, we add a term based on the order :math:`\lambda` spherical expansion:
+#
+# .. math::
+#   \mathbf{T}_\lambda^{\rm corr}(\{\mathbf{r}_i\}) \;=\;
+#   \sum_{\beta=1}^{2\lambda+1}
+#   h_\beta(\{\mathbf{r}_i\})\,
+#   \sum_{z,n}W^{\rm corr}_{\beta,zn}\,\boldsymbol{\rho}_{z,n}^{l=\lambda}
+#
+# with learnable scalar functions :math:`h_\beta(\{\mathbf{r}_i\})`.
+#
+# 5. Scalar Network (SOAP-BPNN)
+# """""""""""""""""""""""""""""
+#
+# We first compute **SOAP powerspectrum** features:
+#
+# .. math::
+#
+#   p_{z_1z_2,n_1 n_2,l} = \bigl(\boldsymbol{\rho}_{z_1,n_1 l} \widetilde{\otimes}
+#   \boldsymbol{\rho}_{z_2,n_2 l}\bigr)_0
+#
+# and then apply a small, per-species multi-layer perceptron to map these features to
+# scalar coefficients :math:`f,g,h`.
+#
+# 6. Assembly and Global Output
+# """""""""""""""""""""""""""""
+#
+# Finally, we assemble the tensor:
+#
+# .. math::
+#
+#   \mathbf{T}_\lambda(\{\mathbf{r}_i\}) = \sum_{\beta=1}^{2\lambda+1}
+#   s_\beta(\{\mathbf{r}_i\}) \, \mathbf{B}^\lambda_\beta
+#   (\mathbf{q}_1,\mathbf{q}_2,\mathbf{q}_3) + \mathbf{T}_\lambda^{\mathrm{corr}}
+#   (\{\mathbf{r}_i\}),
+#
+# where :math:`\mathbf{B}^\lambda_\beta` is a shorthand for the basis tensors and
+# :math:`s_\beta` for the scalar coefficients. For global properties we sum over all
+# atoms.
+
+# %%
+# Training and evaluation of the model
+# ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
+#
+# We train the model with the ``metatrain`` package via its command-line interface.
+# To start training, we run
+#
+# .. code-block:: shell
+#    mtt train data/options.yaml
+#
+# The options file specifies the model architecture and the training parameters:
+#
+# .. literalinclude:: data/options.yaml
+#    :language: yaml
+#
+
+# %%
+subprocess.run(
+    [
+        "mtt",
+        "train",
+        "data/options.yaml",
+    ],
+    check=True,
+)
+# %%
+# We visualize training and validation losses as functions of the epoch.
+# The training log is stored in CSV format in the `outputs` directory.
+
+train_log = np.genfromtxt(
+    glob("outputs/*/*/train.csv")[-1],
+    delimiter=",",
+    names=True,
+    dtype=None,
+    encoding="utf-8",
+)[1:]
+
+plt.loglog(train_log["Epoch"], train_log["training_loss"], label="Training")
+plt.loglog(train_log["Epoch"], train_log["validation_loss"], label="Validation")
+plt.xlabel("Epoch")
+plt.ylabel("Loss")
+plt.legend()
+plt.title("Training and validation loss")
+
+# %%
+# We evaluate the model on the test set using the ``metatrain`` command-line interface:
+#
+# .. code-block:: shell
```

Comment:
You need an empty line to start a block in RST - this is why all (most) the code blocks are not show in the rendered page.

Reply:
Note that when you build offline there should be warnings that you can use to hunt these down.